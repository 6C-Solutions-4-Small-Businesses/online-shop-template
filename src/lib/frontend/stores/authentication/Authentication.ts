import type {UserAccountSummaryPresentation} from '$lib/frontend/presentations/UserAccountSummaryPresentation'
import type {ModalSettings, ModalStore} from '@skeletonlabs/skeleton'
import type {FoundUserPresentation} from '$lib/frontend/presentations/FoundUserPresentation'

export async function openAuthenticationModal(
    store: ModalStore
): Promise<UserAccountSummaryPresentation | string | undefined> {
    return await new Promise<UserAccountSummaryPresentation | string | undefined>((resolve) => {
        const modal: ModalSettings = {
            type: 'component',
            title: 'Identification par courriel',
            component: 'identificationModal',
            body: 'Tapez votre adresse courriel afin que nous puissions vous identifier.',
            buttonTextSubmit: 'Soumettre',
            buttonTextCancel: 'Annuler',

            response: (userInfo: FoundUserPresentation | null) => {
                if (userInfo === null || userInfo === undefined) {
                    resolve(undefined)
                    return
                }
                resolve(userInfo.account ?? userInfo.email)
            }
        }
        store.trigger(modal)
    })
}
