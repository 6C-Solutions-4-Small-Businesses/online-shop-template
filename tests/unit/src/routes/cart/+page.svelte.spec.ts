import CartPage from '$routes/cart/+page.svelte'
import '@testing-library/jest-dom'
import {cleanup, render, screen, waitFor} from '@testing-library/svelte'
import type {ModalStore} from '@skeletonlabs/skeleton'
import {mock} from 'vitest-mock-extended'
import type {UserAccountSummaryPresentation} from '$lib/frontend/presentations/UserAccountSummaryPresentation'
import {goto} from '$app/navigation'
import {afterEach, type Mock} from 'vitest'
import {openAuthenticationModal} from '$lib/frontend/stores/authentication/Authentication'

const {cart} = await vi.hoisted(
    () => import('$mocks/src/lib/frontend/stores/ShoppingCartStore')
)

describe('Page at /cart', () => {
    const checkoutButtonTestId = 'checkout-button'

    beforeEach(() => {
        vi.mock('$lib/frontend/stores/shoppingCartStore/ShoppingCartStore', async () => {
            return {
                cart
            }
        })
        vi.mock('@skeletonlabs/skeleton', async () => {
            const actual = await import('@skeletonlabs/skeleton')
            const {modalStore} = await import('$mocks/src/lib/frontend/components/Modals')
            return {
                ...actual,
                getModalStore: (): ModalStore => modalStore
            }
        })
        vi.mock('$app/navigation', async () => {
            return {
                goto: vi.fn()
            }
        })

        render(CartPage)
    })

    describe('Checkout Button', () => {
        describe('behavior', () => {
            it(`should be in the document under testId: "${checkoutButtonTestId}"`, async () => {
                expect(await screen.findByTestId(checkoutButtonTestId)).toBeInTheDocument()
            })

            it(`should be enabled`, async () => {
                expect(await screen.findByTestId(checkoutButtonTestId)).toBeEnabled()
            })
        })

        describe('when clicked', () => {
            const userEmail = 'a@mail.com'
            const existingUserCustomerId = 'cu_Xy1'

            let openAuthenticationModalMock: Mock
            let checkoutButton: HTMLButtonElement

            beforeAll(() => {
                vi.mock('$lib/frontend/stores/authentication/Authentication', async () => {
                    return {
                        openAuthenticationModal: vi.fn()
                    }
                })
            })

            beforeEach(async () => {
                openAuthenticationModalMock = openAuthenticationModal as Mock
                checkoutButton = await screen.findByTestId(checkoutButtonTestId)
            })

            it(`should take the user to "/cart/checkout?customerId=${existingUserCustomerId}" when prompt resolves with an existing user`, async () => {
                openAuthenticationModalMock.mockResolvedValue(
                    mock<UserAccountSummaryPresentation>({
                        email: userEmail,
                        customerId: existingUserCustomerId
                    })
                )

                checkoutButton.click()

                await waitFor(() =>
                    expect(goto).toHaveBeenCalledWith(
                        `/cart/checkout?customerId=${encodeURIComponent(existingUserCustomerId)}`
                    )
                )
            })

            it(`should take the user to "/cart/checkout?email=${userEmail}" when prompt resolves with a non-existing user`, async () => {
                openAuthenticationModalMock.mockResolvedValue(userEmail)

                checkoutButton.click()

                await waitFor(() =>
                    expect(goto).toHaveBeenCalledWith(`/cart/checkout?email=${encodeURIComponent(userEmail)}`)
                )
            })

            it('should take the user to "/cart/checkout?email=${userEmail}&phoneNumber={phoneNumber}" when prompt resolves with a user with an undefined customerId in the account', async () => {
                const phoneNumber = '1234567890'
                openAuthenticationModalMock.mockResolvedValue(mock<UserAccountSummaryPresentation>({
                    email: userEmail,
                    customerId: undefined,
                    phoneNumber,
                }))

                checkoutButton.click()

                await waitFor(() =>
                    expect(goto).toHaveBeenCalledWith(`/cart/checkout?email=${encodeURIComponent(userEmail)}&phoneNumber=${encodeURIComponent(phoneNumber)}`)
                )
            })
        })
    })

    afterEach(() => {
        vi.clearAllMocks()
        cleanup()
    })
})
