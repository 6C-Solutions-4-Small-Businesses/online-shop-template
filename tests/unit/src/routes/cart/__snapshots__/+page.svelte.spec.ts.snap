// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Page at /cart > should have the expected HTML structure 1`] = `
<div>
  <div
    class="w-full max-w-full h-screen bg-slate-100 overflow-y-scroll md:overflow-hidden"
  >
    <div
      class="title text-3xl font-thin pt-5 pl-5"
    >
      Shopping Cart
    </div>
     
    <div
      class="w-full p-5 grid gap-4 sm:grid-cols-2 md:grid-cols-3 lg:grid-cols-4 xl:grid-cols-4"
    >
      
    </div>
  </div>
   
  <div
    class="sticky flex flex-col justify-center items-center gap-5 bottom-0 bg-white h-42 sm:flex-row sm:justify-end sm:items-center sm:h-22 md:h-30"
  >
    <div
      class="text-xl font-semibold text-primary"
    >
      Sub Total: 
      <span
        class="text-28 leading-[33px] font-semibold text-primary"
      >
        $
        NaN
      </span>
    </div>
     
    <button
      class="uppercase w-[300px] h-[39px] mr-6 rounded-[10px] text-white bg-primary font-semibold text-base"
      data-testid="checkout-button"
    >
      checkout
    </button>
  </div>
  
</div>
`;
